### PR TITLE
Let ExtensionPointBinding be garbage collected.

### LINKED ISSUE #97: `bind_extension_point` prevents object garbage collection
`bind_extension_point` stores a reference to the binding and prevents collection of the object, and it does not offer any way to `unbind` either.

`ExtensionPointBinding` objects are stored in a `WeakKeyDictionary` mapping `object` weakref to the `ExtensionPointBinding` instance, but the binding itself stores a strong reference to the `object` and hence prevents its garbage collection.

### PR DESCRIPTION
**Pull request by pankajp**
Fixes #97

`ExtensionPointBinding` now holds only weak references to `obj`
and `extension_registry` so that they all can be garbage collected.
This is needed because of the global registry of
`ExtensionPointBinding` instances held at its class level.

### CODE REVIEW

**Comment by mdickinson** [comment]
I'm not convinced that this is a good way to solve the issue: I'd prefer to see public mechanisms introduced to:

(1) unbind a specific extension point, and
(2) clear out the registry of extension points.

The upshot of #98 is that, by happy accident, we're not actually using weakref mechanics at all. I think that's a _good_ thing: there's too much scope for weakrefs to introduce surprising and non-deterministic behaviour.

File: envisage/extension_point_binding.py [hunk 1]
```diff
@@ -23,15 +23,15 @@ class ExtensionPointBinding(HasTraits):
     #### 'ExtensionPointBinding' interface ####################################
 
     # The object that we are binding the extension point to.
-    obj = Any
+    obj = WeakRef
```
Comment (mdickinson) [inline envisage/extension_point_binding.py]: With this change, `self.obj` might be `None` (and might become `None` at unpredictable moments). So we'd need to update code that uses `self.obj` to allow for that, for example in `_set_trait` and `update_trait`.

File: envisage/extension_point_binding.py [hunk 2]
```diff
@@ -23,15 +23,15 @@ class ExtensionPointBinding(HasTraits):
...
 
     # The Id of the extension point.
     extension_point_id = Str
 
     # The extension registry used by the binding. If this trait is not set then
     # the class-scope extension registry set on the 'ExtensionPoint' class is
     # used (and if that is not set then the binding won't work ;^)
-    extension_registry = Instance(IExtensionRegistry)
+    extension_registry = WeakRef(IExtensionRegistry)
```
Comment (mdickinson) [inline envisage/extension_point_binding.py]: Is this change necessary to fix #97? If all the target objects are already garbage collected, won't the extension registry also have been collected?

Comment (mdickinson) [inline envisage/extension_point_binding.py]: I'm also worried about unintended consequences of this change. For example, if someone creates an extension registry in an existing application, creates some bindings using that registry, and then neglects to store it anywhere, right now there will be no ill effects, since the extension registry is kept alive by the bindings. But with this change, we'd be getting a lot of TraitErrors in that situation. We may see similar effects at application tearDown, when an extension registry has been cleaned up but the bindings trying to use it are still active.

File: envisage/extension_point_binding.py (hunk 1 above)
Comment (mdickinson) [inline envisage/extension_point_binding.py]: Similarly, all uses of `self.extension_registry` need to be examined, and possibly guarded with a `None` check.

**Comment by pankajp** [comment]
@mdickinson I'ave added relevant `None` checks in this PR and also submitted another PR ( #99 ) one which just add api for removing extension point bindings.

**Comment by mdickinson** [comment]
I'm closing this. I don't think it's the right solution to the problem.
